### PR TITLE
Add an option for setup:config:set to define the document root

### LINKED ISSUE #29523: [Issue] Add an option for setup:config:set to define the document root
This issue is automatically created based on existing pull request: magento/magento2#29176: Add an option for setup:config:set to define the document root

---------

### Preconditions: (*)

Add `bin/magento setup:config:set --document-root-is-pub`
I think this is needed in order to raise awareness for https://devdocs.magento.com/guides/v2.3/install-gde/tutorials/change-docroot-to-pub.html section 3, and to simplify automating this setting in deployment scripts.

### Steps to reproduce: (*)

Run in console: `bin/magento setup:config:set --document-root-is-pub` **1** or **true** or **yes**

**Expected Result:** :heavy_check_mark: the values ​​should be in the correct format (boolean)

Screenshot from 2020-09-10 16-22-30

Run in console: `bin/magento setup:config:set --allow-parallel-generation` **1** or **true** or **yes**

**Expected Result:** :heavy_check_mark: the values ​​should be in the correct format (boolean)

Screenshot from 2020-09-10 16-26-13

Run in console: `in/magento setup:config:set --consumers-wait-for-messages **0** or **false** or **no**

**Expected Result:** :heavy_check_mark: the values should be in the correct format (1/0)

Screenshot from 2020-09-10 16-27-43

Actual Result:
### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by AntonEvers**
### Description (*)

Add `bin/magento setup:config:set --document-root-is-pub`
I think this is needed in order to raise awareness for https://devdocs.magento.com/guides/v2.3/install-gde/tutorials/change-docroot-to-pub.html section 3, and to simplify automating this setting in deployment scripts.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29523: Add an option for setup:config:set to define the document root

### CODE REVIEW

**Comment by AntonEvers** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@AntonEvers good idea!
Could you cover this new command with any kind of test? I think unit will be ok there.

Based on the description i have few questions:
1. What if we'll run just `bin/magento setup:config:set --document-root-is-pub`? Will it set to "true" or it will ask for setting true / false?
2. What if we'll run just `bin/magento setup:config:set --document-root-is-pub true`? Will it set to "true"?
3. What if we'll run `bin/magento setup:config:set --document-root-is-pub false`? Will it set it as "false"?

Would be nice to have more clear  description for the command option

**Comment by AntonEvers** [comment]
> @AntonEvers good idea!
> Could you cover this new command with any kind of test? I think unit will be ok there.
> 
> Based on the description i have few questions:
> 
> 1. What if we'll run just `bin/magento setup:config:set --document-root-is-pub`? Will it set to "true" or it will ask for setting true / false?
> 2. What if we'll run just `bin/magento setup:config:set --document-root-is-pub true`? Will it set to "true"?
> 3. What if we'll run `bin/magento setup:config:set --document-root-is-pub false`? Will it set it as "false"?
> 
> Would be nice to have more clear description for the command option

Hi @ihor-sviziev 
The command option currently works similar to `--allow-parallel-generation`, `--skip-db-validation`, `--db-ssl-verify`, `--cleanup-database` and `--use-sample-data`.
It is a flag, rather than an option and it does not allow a value. `bin/magento setup:config:set --document-root-is-pub` would add the following to `env.php`:
```php
'directories' => [
    'document_root_is_pub' => true
]
```
I will add some unit testing. Please let me know if the command should behave more like `--use-secure=USE-SECURE` instead, to allow setting it both to `true` or `false`.

**Comment by AntonEvers** [comment]
I changed the command to a boolean select option.
During interactive installation it will now asK:
```
Is the webserver's document root set to ./pub/?
  [0] no
  [1] yes
```
It accepts inputs: yes, no, 1, 0, true, false.

**Comment by AntonEvers** [comment]
Now I have set the default value to false, so that the key is always present in env.php. This should make developers more aware of the fact that this option exists.

**Comment by ihor-sviziev** [comment]
From first look - looks good!
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @AntonEvers ,
Could you fix static tests? Everything else looks good

**Comment by AntonEvers** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            return false;
+        }
+
+        throw new InputException(__('%1 requires a boolean value.', self::INPUT_KEY_DOCUMENT_ROOT_IS_PUB));
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function validate(array $options, DeploymentConfig $deploymentConfig)
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php]: Until it get merged - can we add there return types?
```suggestion
    public function validate(array $options, DeploymentConfig $deploymentConfig): array
```

File: setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ * Deployment configuration options for the document root root
+ */
+class Directories implements ConfigOptionsListInterface
+{
+    public const INPUT_KEY_DOCUMENT_ROOT_IS_PUB = 'document-root-is-pub';
+
+    /**
+     * @inheritdoc
+     */
+    public function getOptions()
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php]: Can we define here return type?
```suggestion
    public function getOptions(): array
```

File: setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                'Is the webserver\'s document root set to ./pub/',
+                'no'
+            ),
+        ];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function createConfig(array $options, DeploymentConfig $deploymentConfig)
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php]: Can we define here return type?

File: setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    protected function setUp(): void
+    {
+        $this->deploymentConfigMock = $this->createMock(DeploymentConfig::class);
+        $this->configOptionsList = new DirectoriesConfigOptionsList();
+    }
+
+    /**
+     * testGetOptions
+     */
+    public function testGetOptions()
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php]:
```suggestion
    public function testGetOptions(): void
```

File: setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        $this->assertCount(1, $options);
+        $this->assertArrayHasKey(0, $options);
+        $this->assertInstanceOf(SelectConfigOption::class, $options[0]);
+        $this->assertEquals('document-root-is-pub', $options[0]->getName());
+    }
+
+    /**
+     * testCreateConfigCacheRedis
+     */
+    public function testCreateConfig()
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php]:
```suggestion
    public function testCreateConfig(): void
```

File: setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            );
+
+            $this->assertEquals($testData['config'], $configData->getData());
+        }
+    }
+
+    /**
+     * testCreateConfigCacheRedis
+     */
+    public function testCreateConfigInvalid()
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php]:
```suggestion
    public function testCreateConfigInvalid(): void
```

File: setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        $this->assertInstanceOf(SelectConfigOption::class, $options[0]);
+        $this->assertEquals('document-root-is-pub', $options[0]->getName());
+    }
+
+    /**
+     * testCreateConfigCacheRedis
+     */
+    public function testCreateConfig()
+    {
+        $tests = [
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php]: Could you move it to data provider? Could you move it to data provider?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Looks like that’s last part of requests changes, sorry for not requesting them earlier.

**Comment by AntonEvers** [comment]
No problem at all @ihor-sviziev!

**Comment by AntonEvers** [comment]
@magento run all tests

**Review by ihor-sviziev — DISMISSED** [review]
✔ Approved.

Failing tests looks not related to changes form this PR.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7889](https://jira.corp.magento.com/browse/ENGCOM-7889) has been created to process this Pull Request

File: setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+use Magento\Framework\App\DeploymentConfig;
+use Magento\Framework\Config\ConfigOptionsListConstants;
+use Magento\Framework\Config\Data\ConfigData;
+use Magento\Framework\Config\File\ConfigFilePool;
+use Magento\Framework\Exception\InputException;
+use Magento\Framework\Setup\ConfigOptionsListInterface;
+use Magento\Framework\Setup\Option\SelectConfigOption;
+
+/**
+ * Deployment configuration options for the document root root
```
Comment (engcom-Alfa) [inline setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php]: @AntonEvers Could you delete the duplicate word - **root**? Thanks!

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi, @AntonEvers . During testing, we faced some issues.

**Case 1.** Unable to set true/false configuration value with **0** and **1**

Run in console: `bin/magento setup:config:set --document-root-is-pub 1` for example;

**Actual Result:** :heavy_multiplication_x: Isn't supported value '1'

Screenshot from 2020-07-28 14-58-52

**Case 2.** Unable to set true/false configuration without value

Run in console: `bin/magento setup:config:set --document-root-is-pub` 

**Actual Result:** :heavy_multiplication_x: The option requires a value

Screenshot from 2020-07-28 14-26-53

Although, if we run the command like `bin/magento setup:config:set --allow-parallel-generation` for example, :heavy_check_mark:  we can choose **Y/n**

Screenshot from 2020-07-28 15-03-57

@AntonEvers Could you take a look?

Thanks!

**Comment by AntonEvers** [comment]
@engcom-Alfa I would like to connect this PR with https://github.com/magento/magento2/pull/29187 in order to fix the 0/1 issue. Can you please help me figure out what is going on with the tests there?

**Comment by AntonEvers** [comment]
@magento combine 29187 29176

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@ihor-sviziev @engcom-Alfa something in the addition of a BooleanConfigOption is messing up all of the tests, most likely during the setup commands that are part of the test, but I have no insight in that. Can you help me figure this one out?

**Comment by ihor-sviziev** [comment]
@magento combine 29187

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev. Thank you for your request. I'm working on combining the pull requests for you

**Comment by magento-engcom-team** [comment]
@AntonEvers all pull requests have been successfully combined together: 
 - merged https://github.com/magento/magento2/pull/29187: add boolean selector for interactive installations

**Comment by ihor-sviziev** [comment]
@slavvka @lenaorobei could you add some details why tests fails? Seems like we have some fatal error?

**Comment by lenaorobei** [comment]
Yes, there is an error during installation.

```
[2020-07-31 09:23:26] report.ERROR: There are no commands defined in the "deploy:mode" namespace.
#0 /var/www/html/vendor/symfony/console/Application.php(650): Symfony\Component\Console\Application->findNamespace()
#1 /var/www/html/vendor/symfony/console/Application.php(235): Symfony\Component\Console\Application->find()
#2 /var/www/html/lib/internal/Magento/Framework/Console/Cli.php(115): Symfony\Component\Console\Application->doRun()
#3 /var/www/html/vendor/symfony/console/Application.php(147): Magento\Framework\Console\Cli->doRun()
#4 /var/www/html/bin/magento(23): Symfony\Component\Console\Application->run()
#5 {main} [] []
```

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run Static Tests, Unit Tests, Integration Tests, Functional Tests B2B

**Comment by AntonEvers** [comment]
@magento run Integration Tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@ihor-sviziev Static and Functional test fails are not caused by this PR.

**Comment by ihor-sviziev** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
@magento run Static Tests

**Comment by AntonEvers** [comment]
@magento run Static Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7889](https://jira.corp.magento.com/browse/ENGCOM-7889) has been created to process this Pull Request

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @AntonEvers . During testing, we faced some issues.

**Case 1.** Unable to set value for `--consumers-wait-for-messages`

Run in console: `bin/magento setup:config:set --consumers-wait-for-messages=0` (or **no** or **false**)

**Actual Result:** :heavy_multiplication_x: The value does not change



The same situation with enabling option.

**Case 2.** The value for the `--allow-parallel-generation` option is stored in the wrong format

Run in console: `bin/magento setup:config:set --allow-parallel-generation 1` or **no** or **false**

**Actual Result:** :heavy_multiplication_x: All values are stored in `' '`





According to the Magento [documentation](https://devdocs.magento.com/guides/v2.4/config-guide/redis/redis-pg-cache.html) should be as follows:
Screenshot from 2020-08-10 13-47-39

@AntonEvers Could you take a look?

Thanks!

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento create issue

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7889](https://jira.corp.magento.com/browse/ENGCOM-7889) has been created to process this Pull Request

File: setup/src/Magento/Setup/Model/ConfigOptionsList/Cache.php [hunk 9]
```diff
@@ -293,6 +299,19 @@ private function getDefaultConfigValue($inputKey)
...
+     * Convert any valid input option to a boolean
+     *
+     * @param mixed $option
+     *
+     * @return int|null
+     */
+    private function boolVal($option): ?int
+    {
+        return (int)in_array(strtolower((string)$option), BooleanConfigOption::OPTIONS_POSITIVE);
+    }
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Model/ConfigOptionsList/Cache.php]:
```suggestion
    /**
     * Convert any valid input option to a boolean
     *
     * @param mixed $option
     *
     * @return bool
     */
    private function boolVal($option): bool
    {
        return in_array(strtolower((string)$option), BooleanConfigOption::OPTIONS_POSITIVE);
    }
```

Accodring to https://github.com/magento/magento2/blob/5daadaee54d523c9f6bb3939bcb2df70539d8795/lib/internal/Magento/Framework/Cache/LockGuardedCacheLoader.php#L121-L127 and 
https://github.com/magento/magento2/blob/5daadaee54d523c9f6bb3939bcb2df70539d8795/lib/internal/Magento/Framework/Cache/LockGuardedCacheLoader.php#L142-L144

this value should be always boolean

File: setup/src/Magento/Setup/Model/ConfigOptionsList/Directory.php [hunk 10]
```diff
@@ -8,83 +8,71 @@
...
-            $configData->set(
-                self::CONFIG_PATH_DOCUMENT_ROOT_IS_PUB,
-                \filter_var($options[self::INPUT_KEY_DOCUMENT_ROOT_IS_PUB], FILTER_VALIDATE_BOOLEAN)
-            );
+
+        foreach ($deploymentOption as $inputKey => $configPath) {
+            if (!isset($options[$inputKey])) {
+                continue;
+            }
+            $configData->set($configPath, (int)$this->boolVal($options[$inputKey]));
```
Comment (ihor-sviziev) [inline setup/src/Magento/Setup/Model/ConfigOptionsList/Directory.php]: I'm not sure how unit test not failing, but seems like it should be like that:
```suggestion
            $configData->set($configPath, $this->boolVal($options[$inputKey]));
```

Could you double check the test?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@AntonEvers 
I got feedback from @engcom-Alfa that it still setting 1/0 for the `allow_parallel_generation` and `document_is_root`, while docs saying they have to be boolean.

Could you review my comments?
https://github.com/magento/magento2/pull/29176#pullrequestreview-471464208
https://github.com/magento/magento2/pull/29176#pullrequestreview-471464715

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@AntonEvers I just fixed listed above issues. 
FYI assertSame correctly checks all types, while assertEquals things that '0' and false are the same.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7889](https://jira.corp.magento.com/browse/ENGCOM-7889) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

Run in console: `bin/magento setup:config:set --document-root-is-pub` **1** or **true** or **yes**

**After:** :heavy_check_mark: the values ​​are written in the correct format (boolean)

Screenshot from 2020-09-10 16-22-30

Run in console: `bin/magento setup:config:set --allow-parallel-generation` **1** or **true** or **yes**

**After:** :heavy_check_mark: the values ​​are written in the correct format (boolean)

Screenshot from 2020-09-10 16-26-13

Run in console: `in/magento setup:config:set --consumers-wait-for-messages **0** or **false** or **no**


**After:** :heavy_check_mark: the values ​​are written in the correct format (1/0)

Screenshot from 2020-09-10 16-27-43

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@gabrieldagama @sidolov is there any reasons why this PR isn't getting merged?

**Comment by gabrieldagama** [comment]
@ihor-sviziev We are having some issues with its tests that are being looked at internally. Just for reference, the internal ticket is https://jira.corp.magento.com/browse/MTS-1593

**Comment by ihor-sviziev** [comment]
Hi @AntonEvers,
According to the recent changes in 2.4-develop - the support of non-pub was removed and now Magento always uses the pub directory.
https://github.com/magento/magento2/commit/640cad53009b291334234ccd61ab79f256b43da2#diff-2323d6157444b3cf24d4d76635f27185a4e6ad979b0ec2af53e6592d3ca465d5R13

Due to this change I think we could close this PR as not relevant anymore.
 What do you thing?

**Comment by hostep** [comment]
@ihor-sviziev: nice find!

But I really don't understand how they can make such a major change in a minor release? This change will probably be scheduled for Magento 2.4.2, even though it should be for 2.5.0. This might break a lot of installations.

I'm not saying it's a bad thing, it should have happened much earlier, but big changes like these really shouldn't be done in a minor release.

(sorry to interrupt here in this PR)

**Comment by ihor-sviziev** [comment]
@sivaschenko @sidolov could you provide more info why this commit was merged to 2.4-develop as it's introducing backward incompatible changes? https://github.com/magento/magento2/commit/640cad53009b291334234ccd61ab79f256b43da2#diff-2323d6157444b3cf24d4d76635f27185a4e6ad979b0ec2af53e6592d3ca465d5R13

**Comment by shiftedreality** [comment]
@hostep we updated Apache config to use pub directory. Since Apache reads configs on-fly, we didn't notice any issue having it updated.

This solves a few issues:
- Deprecates usage of `\Magento\Config\Model\Config\Reader\Source\Deployed\DocumentRoot`. This class was added but not used in the places where it should.
- Standardizes pub directory paths all over magento without class above
- Standartazise both Apache and Nginx entrypoints

Without this change, we saw a lot of issues while adding RemoteStorage functionality and saw possible issues for developers who work with FS API.

Could you provide any possible risks which we might miss?

**Comment by hostep** [comment]
@shiftedreality: I understand why this is a good improvement, it really is, I and many others have been wanting it even [2 years before](https://github.com/magento/magento2/issues/372#issuecomment-25754683) Magento 2.0.0 was launched.

But where I'm seeing problems is that certain users now will be surprised when they perform a minor update and notice they have to change their document root and somehow can't. I'm pretty sure that there are (cheap) hosting companies around which don't allow you to change the document root and make it very difficult to host magento from inside the `pub` directory. Which is a shame and people shouldn't use these sort of hosting companies, but yes, a lot of people want cheap hosting... So I think it only makes sense to introduce such a big change in a major release like 2.5.0 to avoid breaking shops just because they made a wrong decision in choosing a certain hosting company.
If that's really not possible and it needs to be shipped in 2.4.x, then the release notes will have to add a very big warning on top (and not hidden in the hundreds of changes) so that users are aware of this change.

**Comment by shiftedreality** [comment]
@hostep my understanding is those Customers have an Apache which reads `.htaccess` file. So, it should be supported fine, since the file is updated. Isn't this standard case for Apache on hosting companies?

**Comment by hostep** [comment]
The `.htaccess` file in the Magento root is deleted, so I'm not sure what you are saying?
Also, you can not change the document root from a `.htaccess` file as far as I'm aware.

**Comment by hostep** [comment]
Aaaaaah ok, the github interface tricked me, I assumed it was removed when I saw all those red blocks 😄 
Okay, thanks for the explanation, those new lines in there will probably solve it (untested), sorry about the noise! 🙂

**Comment by shiftedreality** [comment]
Sure, thank you for paying so much attention and passion for Magento!

**Comment by ihor-sviziev** [comment]
@shiftedreality thx for good explanation!

**Pull request closed without merging**